realhf/impl/model/modules/moe/experts.py

```diff
@@ -0,0 +1,230 @@
+        self.local_experts = torch.nn.ModuleList()
+
+        for _ in range(self.num_experts):
+            expert = LlamaLayerNormMLP(
+                hidden_dim=config.hidden_dim,
+                intermediate_dim=config.intermediate_dim,
+                activation_function=config.activation_function,
+                is_expert=True,
+                model_parallel=constants.model_parallel_world_size() > 1,
+                gradient_accumulation_fusion=config.gradient_accumulation_fusion,
+                dtype=dtype,
+                device=device,
+            )
+            self.local_experts.append(expert)
+
+    def forward(
+        self,
+        permuted_local_hidden_states: torch.Tensor,
+        tokens_per_expert: torch.Tensor,
+    ):
+        output_local = torch.zeros_like(permuted_local_hidden_states)
+        cumsum_num_tokens = torch.cumsum(tokens_per_expert, dim=0)
+        # Insert zero at the begining for offset index's convenience
+        zero_tensor = torch.zeros(1, dtype=torch.long, device=cumsum_num_tokens.device)
+        cumsum_num_tokens = torch.cat((zero_tensor, cumsum_num_tokens))
+
+        for expert_num, expert in enumerate(self.local_experts):
+            start = cumsum_num_tokens[expert_num]
+            end = cumsum_num_tokens[expert_num + 1]
+            hidden = permuted_local_hidden_states[start:end]
+            output = expert(hidden)
+            output_local[start:end] = output
+
+        return output_local
+
+
+class ExpertParam(torch.nn.Module):
+    """A dummy class that maps weight tensors in GroupedMLP to pytorch parameters
+    for compatibility of weight saving/loading.
+    """
+
+    def __init__(
+        self,
+        gate_proj: torch.Tensor,
+        up_proj: torch.Tensor,
+        down_proj: torch.Tensor,
+    ):
+        class LinearParam(torch.nn.Module):
+            def __init__(self, param: torch.Tensor):
+                super(LinearParam, self).__init__()
+                self.weight = Parameter(param)
+
+        super(ExpertParam, self).__init__()
+
+        self.gate_proj = LinearParam(gate_proj)
+        self.up_proj = LinearParam(up_proj)
+        self.down_proj = LinearParam(down_proj)
+
+
+class GroupedMLP(torch.nn.Module):
+    """An efficient implementation of the Experts layer using CUTLASS GroupedGEMM.
+    See https://github.com/tgale96/grouped_gemm for details.
+
+    This class is designed to execute multiple experts in parallel, thereby maximizing computational efficiency.
+    """
+
+    def __init__(
+        self,
+        config: ReaLModelConfig,
+        init_method: Callable = init.xavier_normal_,
+        dtype: Optional[torch.dtype] = None,
+        device: Optional[Union[str, torch.device]] = None,
+    ):
+        super().__init__()
+        assert (
+            not constants.sequence_parallel()
+        ), "Grouped GEMM does not support sequence parallel"
+        assert dtype == torch.bfloat16, "Grouped GEMM only supports bfloat16 dtype"
```

Comment:
Input dtype None will create fp32 weights, but this assersion will not work with dtype == None.

Reply:
I did not quite understand the sentence, but this assertion is removed so maybe it does not matter

Reply:
My bad. I don't have a clear mind today.